### PR TITLE
Refactor/evaluation task

### PR DESCRIPTION
**Pull request by AlessandroPolidori**
## Summary

Describe the purpose of the pull request, including:

- What problem does it solve?
Only SegmentationEvaluation extends Evaluation. Also, we need to uniform the evaluation tasks as much as possible. 
Currently they are unnecessarily different (attributes names are different, some use experiment_path while others the model_path).

Also SegmentationEvaluation was not tested, i added two tests (one for multiclass, one for binary).

- How does it solve the problem?
All the children evaluations extend the Evaluation class.

- What are the benefits of the solution?
Clean code.
- Reference any relevant issues or pull requests.
https://github.com/orobix/quadra/issues/21

## Type of Change

Please select the one relevant option below:

- Breaking change (fix or feature that would cause existing functionality to not work as expected

## Checklist

Please confirm that the following tasks have been completed:

- [x] I have tested my changes locally and they work as expected. (Please describe the tests you performed.)
- [x] I have added unit tests for my changes, or updated existing tests if necessary.
- [x] I have updated the documentation, if applicable.
- [x] I have installed pre-commit and run locally for my code changes.

### CODE REVIEW

File: quadra/configs/experiment/base/segmentation/smp_multiclass_evaluation.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  - override /datamodule: base/segmentation_multiclass
+  - override /transforms: default_resize
+  - override /task: segmentation_evaluation
+
+core:
+  tag: "run"
+  name: "quadra_default"
+  upload_artifacts: True
+
+trainer:
```
Comment (lorenzomammana) [inline quadra/configs/experiment/base/segmentation/smp_multiclass_evaluation.yaml:13]: Why the evaluation config has reference to training information :D?

File: quadra/tasks/base.py [hunk 2]
```diff
@@ -290,46 +292,68 @@ def execute(self) -> None:
...
-            log.warning("Report folder is not provided, using default report folder")
-            self.report_folder = "report"
+        self.model_info_filename = "model.json"
+        self.report_path = ""
+        self.metadata = {"report_files": []}
 
     @property
-    def deployment_model(self) -> RecursiveScriptModule:
-        """RecursiveScriptModule: The deployment model."""
+    def deployment_model(self):
```
Comment (lorenzomammana) [inline quadra/tasks/base.py:328]: No typing?

File: quadra/tasks/classification.py [hunk 3]
```diff
@@ -981,29 +962,9 @@ def deployment_model(self, model_path: str):
...
-            log.warning(
-                f"Input width of the model ({self.model_data['input_size'][1]}) is different from the one specified "
-                + f"in the config ({self.config.transforms.input_width}). Fixing the config."
-            )
-            self.config.transforms.input_width = self.model_data["input_size"][1]
-
-        self.deployment_model = self.model_path
-
         super().prepare()
+        self.deployment_model = self.model_path
```
Comment (lorenzomammana) [inline quadra/tasks/classification.py:966]: Why this is not done in the super? We need to manually assign everytime or I'm missing something

Comment (lorenzomammana) [reply quadra/tasks/classification.py:966]: Ok I saw it may be different for plain torch import

File: quadra/tasks/segmentation.py [hunk 4]
```diff
@@ -222,18 +222,22 @@ def __init__(
         self,
         config: DictConfig,
         model_path: str,
-        report_folder: Optional[str] = None,
+        device: str = "cpu",
```
Comment (lorenzomammana) [inline quadra/tasks/segmentation.py:225]: We say in the docstring that device may be None, but here is mandatory

File: quadra/tasks/segmentation.py [hunk 5]
```diff
@@ -222,18 +222,22 @@ def __init__(
...
 
     def save_config(self) -> None:
         """Skip saving the config."""
 
+    def prepare(self) -> None:
+        """Prepare the evaluation."""
+        super().prepare()
+        self.datamodule = self.config.datamodule
+        # prepare_data() must be explicitly called because there is no training
+        self.datamodule.prepare_data()
```
Comment (lorenzomammana) [inline quadra/tasks/segmentation.py:238]: Segmentation datamodule has no check for train or test?

File: quadra/tasks/segmentation.py [hunk 6]
```diff
@@ -269,18 +273,18 @@ class SegmentationAnalysisEvaluation(SegmentationEvaluation):
...
+        device: Device to use for evaluation. If None, the device is automatically determined.
     """
 
-    def __init__(self, config: DictConfig, model_path: str, report_folder: str = "analysis_report"):
-        super().__init__(config=config, model_path=model_path, report_folder=report_folder)
+    def __init__(
+        self,
+        config: DictConfig,
+        model_path: str,
+        device: str = "cpu",
```
Comment (lorenzomammana) [inline quadra/tasks/segmentation.py:283]: Same I think this should be optional

**Review by lorenzomammana — CHANGES_REQUESTED** [review]
Overall looks pretty good to me and a good improvement in the overall code structure

**Comment by rcmalli** [comment]
Amazing job @AlessandroPolidori!

File: quadra/tasks/segmentation.py (hunk 5 above)
Comment (AlessandroPolidori) [reply quadra/tasks/segmentation.py:238]: I'm having trouble grasping the meaning behind your question. What i see from the other tasks is that whenever you have no lightning training you have to manually call the prepare_data(). I added some comments and TODO, se my commit "todo: Add todo for the prepare_data() manual call"

Comment (lorenzomammana) [reply quadra/tasks/segmentation.py:238]: Nada, never review PR at 18:30, I was thinking about setup() :D

File: quadra/tasks/classification.py (hunk 3 above)
Comment (AlessandroPolidori) [reply quadra/tasks/classification.py:967]: Actually the only "problem" exists for patchsklearn and sklearnclassif. which don't have a deployment model but backbone + classifier. To move the self.deployment_model = self.model_path in Evaluation i will overwrite the deployment model setter and set classifier and backbone in there. Without really seting a deployment_model object, which is not used in patchsklearn and sklearnclassification.

Comment (AlessandroPolidori) [reply quadra/tasks/classification.py]: Done, see if you like that. Commit: style: Move self.deployment_model = self.model_path to parent class
p.s. i know it's not "style", "refactor" would have been better

File: quadra/tasks/classification.py [hunk 7]
```diff
@@ -775,26 +769,34 @@ def prepare(self) -> None:
...
+    @property
+    def deployment_model(self):
+        """Deployment model."""
+        return None
+
+    @deployment_model.setter
+    def deployment_model(self, model_path: str):
+        """Set backbone and classifier."""
         try:
             self.backbone = OmegaConf.load(
```
Comment (lorenzomammana) [inline quadra/tasks/classification.py:791]: I don't get it, why should we be forced to use pytorch here? 
I think we should support any kind of model as per other tasks, if torch is not used we don't generate gradients.

Or am I missing something?

File: quadra/tasks/patch.py [hunk 8]
```diff
@@ -341,6 +330,18 @@ def test(self) -> None:
             self.idx_to_class[i] if i != -1 else "N/A" for i in res["real_label"].unique().tolist()
         ]
 
+    @property
+    def deployment_model(self):
+        """Deployment model."""
+        return None
+
+    @deployment_model.setter
+    def deployment_model(self, model_path: str):
```
Comment (lorenzomammana) [inline quadra/tasks/patch.py:339]: Same here, I would probably expect that we can deploy the model without a backbone, but if a deployed model is present it should be the first option

**Review by lorenzomammana — CHANGES_REQUESTED** [review]
I think we need to unify the deployment model handling also for sklearn with the caveat of being able to load a backbone directly if it wasn't exported

File: quadra/tasks/classification.py (hunk 7 above)
Comment (AlessandroPolidori) [reply quadra/tasks/classification.py]: See commit "fix: Model import in sklearn evaluations now supports .pt and .pth model files" and tell me what you think

**Comment by rcmalli** [comment]
Can we also have a short tutorial starting with training a model and running `Evaluation task` from command-line or using Python?
